### DISCUSSION TITLE
Running DNS-proxy alongside tunnel

### DISCUSSION THREAD

**Discussion by dunxd** [homeassistant-apps/app-cloudflared#265] [Ideas]
cloudflared also supports running a DNS over HTTPS proxy that can be used by clients on the network.

Currently I am running this on the Raspberry Pi where I run PiHole, using [the details on the PiHole documentation site.](https://docs.pi-hole.net/guides/dns/cloudflared/). This executues cloudflared using [the proxy-dns subcommand](https://fig.io/manual/cloudflared/proxy-dns).

Since the Pi I am running PiHole on is a model 1B, and there have been segmentation fault issues with cloudflared on it, I wondered whether it would be feasible to run the cloudflared proxy-dns functions in the cloudflared add-on running on my higher capacity Home Assistant server.

As it stands the Add-On doesn't support this, but it might a useful feature for some users in my position or that otherwise want to proxy DNS over HTTPS with Cloudflare or other services.

I note that [the tunnel subcommand also supports DNS related options](https://fig.io/manual/cloudflared/tunnel).

**Comment by brenner-tobias** [comment]
I generally like to idea of using the tunnel for DNS queries to Cloudflare DNS as well. I was looking into configuration option for the Cloudflare managed tunnel, but as far as I can tell, there is no possibility to configure this from the Cloudflare Dashboard.

So we would need to implement an add-on option to set if you want to use the DNS proxy. 

Before looking into it more technically, I would love to hear your thoughts @elcajon.

**Reply by elcajon** [reply]
I am not sure if this fits into this add-on as it implements a DNS server in parallel to the tunnel feature. Tunnel and dns-proxy are both included in cloudflared but the proxy is not part of the tunnel (and therefore cannot be activated from Cloudflare managed tunnel dashboard).

I am also not sure why to use the encrypted tunnel to query already encrypted DNS-over-HTTPS requests?
If you want to use encrypted DNS for a local DNS server (which this feature request is all about as far as I can tell(?)) it would be much easier to use the hassio-addons Adguard add-on with the Cloudflare upstream servers:

![image](https://user-images.githubusercontent.com/52052494/206148035-6c91de72-8ceb-44a3-8f10-7efd96f3f25b.png)

If we want to include this feature we would need to add (at least) an config option and port forwarding for the DNS server to listen on.

**Comment by dunxd** [comment]
Pihole doesn't support DNS over HTTPS. The DNS proxy accepts unencrypted
DNS requests and encrypts them to the upstream server. This would work for
anyone that wanted to update to DNS over HTTPS for their whole network but
didn't want to set up adguard.

The method for configuring mentioned on the pihole documentation sets up a
dedicated proxy-dns, but the [tunnel command](https://fig.io/manual/cloudflared/tunnel) for cloudflared _also_ has
proxy-dns options (e.g. `--proxy-dns` to activate it), so presumably this can be set up alongside tunnel.

**Comment by brenner-tobias** [comment]
Mhm I am unsure how to move forward here. On one hand, the DNS-proxy is a "function" of Cloudflared and we could easily implement it. If we define the add-on to be the HA Cloudflared add-on, it does make sense to offer all options, that the Cloudflare demon offers.
On the other hand it is true that there are plenty options to use DNS over https without this proxy. For most users, it should be possible to do so in your router. If not, you can use other add-ons, like adguard. In any way, we will not promote this feature anywhere and it is unlikely it will be used often, plus it would increase the complexity of the add-on and introduce many new ways of "failing", so maintaining it will become more difficult.

So for now, I suggest we wait and see if there are others here that want to see this feature. If we realize this is needed by many, we can reconsider.
